File: src/visual-graph/prune/collect-ids.test.ts
import { describe, expect, test } from "vitest";

import { DIRECTION } from "../direction.js";
import type {
  Direction,
  VisualElement,
  VisualNode,
  VisualSubgraph,
} from "../model.js";
import { NODE_KIND } from "../node-kind.js";
import { SUBGRAPH_KIND } from "../subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import { collectIds } from "./collect-ids.js";

const node = (id: string): VisualNode => ({
  type: VISUAL_ELEMENT_TYPE.Node,
  id,
  kind: NODE_KIND.Variable,
  name: id,
  line: 1,
  isJsxElement: false,
  endLine: null,
  unused: false,
  declarationKind: null,
  initIsFunction: false,
});

const sg = (
  id: string,
  elements: VisualElement[],
  direction: Direction = DIRECTION.TB,
): VisualSubgraph => ({
  type: VISUAL_ELEMENT_TYPE.Subgraph,
  id,
  kind: SUBGRAPH_KIND.Function,
  line: 1,
  direction,
  elements,
  endLine: null,
  ownerNodeId: "n_owner",
  ownerName: "owner",
});

describe("collectIds", () => {
  test("includes both node ids and subgraph ids (unlike collectNodeIds)", () => {
    const ids = collectIds([
      sg("outer", [node("x"), sg("inner", [node("y")])]),
      node("top"),
    ]);
    expect([...ids].sort()).toEqual(["inner", "outer", "top", "x", "y"]);
  });

  test("empty input → empty set", () => {
    expect(collectIds([]).size).toBe(0);
  });

  test("returns a Set (deduped already by Map semantics)", () => {
    expect(collectIds([node("a")])).toBeInstanceOf(Set);
  });
});

File: crates/unsnarl-visual-graph/src/prune/collect_ids_test.rs
use super::*;

use crate::prune::prune_fixtures::{const_binding_node, function_subgraph};

#[test]
fn includes_both_node_and_subgraph_ids() {
    let ids = collect_ids(&[
        VisualElement::Subgraph(function_subgraph(
            "outer",
            1,
            vec![
                VisualElement::Node(const_binding_node("x", "x", 1)),
                VisualElement::Subgraph(function_subgraph(
                    "inner",
                    1,
                    vec![VisualElement::Node(const_binding_node("y", "y", 1))],
                )),
            ],
        )),
        VisualElement::Node(const_binding_node("top", "top", 1)),
    ]);
    let mut sorted: Vec<String> = ids.into_iter().collect();
    sorted.sort();
    assert_eq!(sorted, vec!["inner", "outer", "top", "x", "y"]);
}

#[test]
fn empty_input_yields_empty_set() {
    assert_eq!(collect_ids(&[]).len(), 0);
}
